### PR TITLE
Update get-codeowners PowerShell file

### PR DESCRIPTION
**Pull request by chidozieononiwu**
Extend CodeOwners.ps1 to recognize Labels and team aliases in code owners file.

Addresses https://github.com/Azure/azure-sdk-for-net/issues/14760 and https://github.com/Azure/azure-sdk-tools/issues/954

Blocked on https://github.com/Azure/azure-sdk-tools/pull/1052

### CODE REVIEW

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [inline eng/common/scripts/get-codeowners.ps1:5]: I wouldn't have guess these are output variable names based on the names. We should name them so that is a little more obvious.

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [inline eng/common/scripts/get-codeowners.ps1:19]: What is this here for?

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [inline eng/common/scripts/get-codeowners.ps1:31]: This assumes exact spacing between `#` and PRlabel we should be a little more flexible. Make sure you have a look at https://github.com/Azure/azure-sdk-tools/blob/f9a1eef96d4b878a60e0dfc687b6f5067ccc2562/tools/CreateRuleFabricBot/CreateRuleFabricBot/Rules/PullRequestLabel/CodeOwnerEntry.cs#L51 and https://github.com/Azure/azure-sdk-tools/blob/11713539bb97ccb0342db98c7c400d31233a42d5/tools/CreateRuleFabricBot/CreateRuleFabricBot/Rules/PullRequestLabel/PullRequestLabelFolderCapability.cs#L159 and try to mimic the same behavior in PS.

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [inline eng/common/scripts/get-codeowners.ps1:66]: That doesn't seem like a save assumption. Why wouldn't we query for the teams under our org?

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [inline eng/common/scripts/get-codeowners.ps1:53]: Do we need an auth token to query the users?

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [reply eng/common/scripts/get-codeowners.ps1:66]: What do plan to do with the teams?

File: eng/common/scripts/get-codeowners.ps1
Comment (chidozieononiwu) [reply eng/common/scripts/get-codeowners.ps1:53]: Not necessary but using authToken increases the rate limit.

File: eng/common/scripts/get-codeowners.ps1
Comment (chidozieononiwu) [reply eng/common/scripts/get-codeowners.ps1:66]: We were planing to auto tag PR with the team alias

Comment (weshaggard) [reply eng/common/scripts/get-codeowners.ps1:66]: auto-tagging the PR's in the docs org will not work as the teams are not necessary the same in each org and the team cannot be accessed from another org.

Comment (chidozieononiwu) [reply eng/common/scripts/get-codeowners.ps1:66]: Am actually doing this to tag automatic version increment PRs.

File: eng/common/scripts/get-codeowners.ps1
Comment (chidozieononiwu) [reply eng/common/scripts/get-codeowners.ps1:19]: Am using it to capture the the line before the current line in the loop. i.e. (current Loop count -1)

File: eng/common/scripts/get-codeowners.ps1
Comment (chidozieononiwu) [reply eng/common/scripts/get-codeowners.ps1]: I updated the Logic.

**Comment by dbradish-microsoft** [comment]
The PowerShell doc lead is @Mike Robbins (Mike.Robbins@microsoft.com)<mailto:Mike.Robbins@microsoft.com>

Kindly,

Delora

**Review by mikefrobbins — CHANGES_REQUESTED** [review]
As a PowerShell best practice:
•	Avoid aliases and positional parameters. Use full cmdlet and parameter names.
•	Try/Catch doesn’t catch non-terminating errors unless you specify “-ErrorAction Stop”.
•	Avoid Write-Host because it doesn’t write to the pipeline.
•	Avoid Hungarian notation.

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [reply eng/common/scripts/get-codeowners.ps1:19]: You don't have to declare the variable here.

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [reply eng/common/scripts/get-codeowners.ps1]: For automatic version increment PR's we shouldn't need to do any of this work. We already have the codeowners being added as reviewers by github and we have the labels being added by the area via the fabric bot. I thought you were doing this for doc PRs.

Comment (weshaggard) [reply eng/common/scripts/get-codeowners.ps1]: Can you add a link to the issue you are trying to address in the PR description?

Comment (chidozieononiwu) [reply eng/common/scripts/get-codeowners.ps1]: I guess you are right. Looks like there is some confusion. This is the [original issue](https://github.com/Azure/azure-sdk-for-net/issues/14760) I set out to fix. But @scbedd requested reviews on my PRs ([python](https://github.com/Azure/azure-sdk-for-python/pull/13485), [java](https://github.com/Azure/azure-sdk-for-java/pull/14693), [net](https://github.com/Azure/azure-sdk-for-net/pull/14767)), he then opened [another issue](https://github.com/Azure/azure-sdk-tools/issues/954). I was attempting to get it all resolved.

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [inline eng/common/scripts/get-codeowners.ps1:67]: Listing these variables doesn't do anything.

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [inline eng/common/scripts/get-codeowners.ps1:75]: I recommend keeping $users and $teams as an array and then doing a join on the values after this loop.

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [inline eng/common/scripts/get-codeowners.ps1:96]: what if there are now users?

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [reply eng/common/scripts/get-codeowners.ps1]: @chidozieononiwu OK so all that was really needed was a way to ignore teams then correct?  I'm fine with adding the other functionality to the script but I don't believe it is going to really be used anywhere yet.

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [inline eng/common/scripts/get-codeowners.ps1:7]: Labels parameter is still here. BTW I was fine with you keeping the labels functionality here as that might become useful at some point as well.

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [inline eng/common/scripts/get-codeowners.ps1:94]:
```suggestion
    Write-Host "##vso[task.setvariable variable=$VsoOwningUsers;]$usersString"
```

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [inline eng/common/scripts/get-codeowners.ps1:92]: We should only do the += if $usersString isn't empty.

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [inline eng/common/scripts/get-codeowners.ps1:102]:
```suggestion
    Write-Host "##vso[task.setvariable variable=$VsoOwningTeams;]$teamsString"
```

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [inline eng/common/scripts/get-codeowners.ps1:100]: You should only do the += if $teamsString isn't empty.

File: eng/common/pipelines/templates/steps/get-pr-owners.yml
Comment (chidozieononiwu) [inline eng/common/pipelines/templates/steps/get-pr-owners.yml:2]: leaft this variable to prevent a break when this PR is merged.

Comment (weshaggard) [reply eng/common/pipelines/templates/steps/get-pr-owners.yml:2]: Great lets file a follow-up to remove it once we convert all usages.

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [inline eng/common/scripts/get-codeowners.ps1:78]: Not sure how much it matters but it seems like we should check this before verifying if it is a user or not. I don't expect users names can have a "/" so this would save a rest call for all teams if we checked the string first.

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [inline eng/common/scripts/get-codeowners.ps1:67]: What explicitly make this an arraylist as opposed to just an array? `$users = @()` and `$users += $str`?

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [inline eng/common/scripts/get-codeowners.ps1:93]: Did you intend to put a space after the comma? I'd probably remove any spaces.

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [inline eng/common/scripts/get-codeowners.ps1:100]:
```
$users += $presendingOwners -split ","
Write-Host "##vso[task.setvariable variable=$VsoOwningUsers;]$usersString"

```

Comment (weshaggard) [inline eng/common/scripts/get-codeowners.ps1:100]: I wonder if it might be better to work with continue with the array like such:
```
if ($presentOwningUsers) {
$users += $presentOwningUsers
}
Write-Host "##vso[task.setvariable variable=$VsoOwningUsers;]$($users -join ',')"
```

File: eng/common/pipelines/templates/steps/get-pr-owners.yml
Comment (chidozieononiwu) [reply eng/common/pipelines/templates/steps/get-pr-owners.yml:2]: https://github.com/Azure/azure-sdk-tools/issues/1035

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [inline eng/common/scripts/get-codeowners.ps1:91]: Should this be a warning?

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [inline eng/common/scripts/get-codeowners.ps1:54]: Lets make remove the join here and keep it an array.

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [inline eng/common/scripts/get-codeowners.ps1:116]: Once $labels is an array just do a += like the users and teams.

**Review by weshaggard — APPROVED** [review]
One clean-up comment otherwise looks good.

File: eng/common/scripts/get-codeowners.ps1
Comment (chidozieononiwu) [reply eng/common/scripts/get-codeowners.ps1:116]: Does not seem like this will work if the returned value from the environment variable is already a comma separated string which I think is what @scbedd was anticipating.

Comment (chidozieononiwu) [reply eng/common/scripts/get-codeowners.ps1:119]: Never mind it'll all be fine when join is called.

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [inline eng/common/scripts/get-codeowners.ps1:105]:
```suggestion
    Write-Host "##vso[task.setvariable variable=$VsoOwningUsers;]$($users -join ',')"
```

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [inline eng/common/scripts/get-codeowners.ps1:113]:
```suggestion
    Write-Host "##vso[task.setvariable variable=$VsoOwningTeams;]$($teams -join ',')"
```

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [inline eng/common/scripts/get-codeowners.ps1:121]:
```suggestion
    Write-Host "##vso[task.setvariable variable=$VsoOwningLabels;]$($labels -join ',')"
```

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [inline eng/common/scripts/get-codeowners.ps1:10]: Any reason not to import common.ps1?

File: eng/common/scripts/get-codeowners.ps1
Comment (weshaggard) [inline eng/common/scripts/get-codeowners.ps1:50]: Given that you split on "," below we should just avoid the join here and just keep an array.

**Comment by ahsonkhan** [comment]
Any updates to this PR? Is it still being worked on?

If this is blocked on something, can you link to the PRs or issues in the description that we are waiting on.

**Comment by chidozieononiwu** [comment]
> Any updates to this PR? Is it still being worked on?
> 
> If this is blocked on something, can you link to the PRs or issues in the description that we are waiting on.

Done

**Comment by chidozieononiwu** [comment]
Closing this, till I get to come back to it.

**Pull request closed without merging**
